### ISSUE TITLE
Issued certificate check fails for some reason

### ISSUE THREAD

**Issue by VojtechMyslivec** [VojtechMyslivec/letsencrypt-zimbra#83]
_A question by @dhcmega riginally posted in https://github.com/VojtechMyslivec/letsencrypt-zimbra/issues/79#issuecomment-2223584422_

---
Hi

I'm trying to figure this out. I tried with the original and the reverted one:

```
#cat "$intermediate_CA_file" "$root_CA_file" > "$chain_file"
cat "$root_CA_file" "$intermediate_CA_file"  > "$chain_file"
```
Neither works:
```
** Verifying '/tmp/tmp.xuHC12YvmQ/0000_cert.pem' against '/opt/zimbra/ssl/zimbra/commercial/commercial.key'
Certificate '/tmp/tmp.xuHC12YvmQ/0000_cert.pem' and private key '/opt/zimbra/ssl/zimbra/commercial/commercial.key' match.
** Verifying '/tmp/tmp.xuHC12YvmQ/0000_cert.pem' against '/tmp/tmp.xuHC12YvmQ/0001_chain.pem'
ERROR: Unable to validate certificate chain: C = US, O = Let's Encrypt, CN = R10
error 2 at 1 depth lookup: unable to get issuer certificate
error /tmp/tmp.xuHC12YvmQ/0000_cert.pem: verification failed
```
On top of this, the currents certs expires tomorrow, great....

8.8.15.GA.4581.FOSS

Thanks!

**Comment by VojtechMyslivec** [comment]
Hi @dhcmega, that seems not to be related with certificate order (#79).


Do you have `letsencrypt-zimbra` updated to the last version? Don't you made any changes? This line from your log output doesn't make sense:
```
** Verifying '/tmp/tmp.xuHC12YvmQ/0000_cert.pem' against '/tmp/tmp.xuHC12YvmQ/0001_chain.pem'
```

It compares issued Letsencrypt' cert `0000_cert.pem` against `0001_chain.pem` which is also output of the `certbot` tool, but `letsencrypt-zimbra` script should prepare `chain.pem`  file with both intermediate and root cert (as Zimbra requires).

The line should look like this:

```
** Verifying '0000_cert.pem' against 'chain.pem'
```

Let me know if you managed to realize what's wrong

**Comment by dhcmega** [comment]
Hi @VojtechMyslivec you were right. I tried that line because the script prints the error:
```
letsencrypt-zimbra.sh: info: start nginx
letsencrypt-zimbra.sh: info: assemble cert files
letsencrypt-zimbra.sh: info: test and deploy certificates
letsencrypt-zimbra.sh: error: Installation of the issued certificate failed.
```
So I didn't know what the problem was and tried that line manually, now that I tried as you suggested, I got this:

```
$ sudo -Hiu zimbra zmcertmgr deploycrt comm /tmp/tmp.usBeqCJNyV/0000_cert.pem /tmp/tmp.usBeqCJNyV/chain.pem
** Verifying '/tmp/tmp.usBeqCJNyV/0000_cert.pem' against '/opt/zimbra/ssl/zimbra/commercial/commercial.key'
Certificate '/tmp/tmp.usBeqCJNyV/0000_cert.pem' and private key '/opt/zimbra/ssl/zimbra/commercial/commercial.key' match.
** Verifying '/tmp/tmp.usBeqCJNyV/0000_cert.pem' against '/tmp/tmp.usBeqCJNyV/chain.pem'
Valid certificate chain: /tmp/tmp.usBeqCJNyV/0000_cert.pem: OK
** Copying '/tmp/tmp.usBeqCJNyV/0000_cert.pem' to '/opt/zimbra/ssl/zimbra/commercial/commercial.crt'
** Copying '/tmp/tmp.usBeqCJNyV/chain.pem' to '/opt/zimbra/ssl/zimbra/commercial/commercial_ca.crt'
** Appending ca chain '/tmp/tmp.usBeqCJNyV/chain.pem' to '/opt/zimbra/ssl/zimbra/commercial/commercial.crt'
** Importing cert '/opt/zimbra/ssl/zimbra/commercial/commercial_ca.crt' as 'zcs-user-commercial_ca' into cacerts '/opt/zimbra/common/lib/jvm/java/lib/security/cacerts'
** NOTE: restart mailboxd to use the imported certificate.
** Saving config key 'zimbraSSLCertificate' via zmprov modifyServer zmail.huayra.com.ar...ok
** Saving config key 'zimbraSSLPrivateKey' via zmprov modifyServer zmail.huayra.com.ar...ok
** Installing imapd certificate '/opt/zimbra/conf/imapd.crt' and key '/opt/zimbra/conf/imapd.key'
** Copying '/opt/zimbra/ssl/zimbra/commercial/commercial.crt' to '/opt/zimbra/conf/imapd.crt'
** Copying '/opt/zimbra/ssl/zimbra/commercial/commercial.key' to '/opt/zimbra/conf/imapd.key'
** Creating file '/opt/zimbra/ssl/zimbra/jetty.pkcs12'
ERROR: openssl pkcs12 export to '/opt/zimbra/ssl/zimbra/jetty.pkcs12' failed(1):
Error creating PKCS12 MAC; no PKCS12KDF support?
Use -nomac if MAC not required and PKCS12KDF support not available.
C0F1DF27407F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:373:Global default library context, Algorithm (PKCS12KDF : 192), Properties (<null>)
C0F1DF27407F0000:error:1180006B:PKCS12 routines:pkcs12_gen_mac:key gen error:crypto/pkcs12/p12_mutl.c:147:
C0F1DF27407F0000:error:1180006D:PKCS12 routines:PKCS12_set_mac:mac generation error:crypto/pkcs12/p12_mutl.c:220:
```

Thanks a lot for taking the time to read all this.

**Comment by defkev** [comment]
> Error creating PKCS12 MAC; no PKCS12KDF support?

Sounds like https://github.com/Zimbra/zm-core-utils/pull/137
Check your OpenSSL version

> openssl version

https://github.com/Zimbra/zm-core-utils/pull/137/files#diff-ca666ebf68794392c2a64d972023d1a9d67cb443a73dbe5ec3abcda645fea3ad

**Comment by VojtechMyslivec** [comment]
Yes, this seems like a problem somewhere in the `zmcertmgr` command itself, not in `letsencrypt-zimbra` script.

FYI: you can call `letsencrypt-zimbra` with `-v` instead of `-q` to get a verbose output next time (see `-h` for help).

**Comment by dhcmega** [comment]
@VojtechMyslivec thank again mate.
Should I modify the script to include the `-fips` flag?
I tried like this, and it did not work:
```
sudo -Hiu zimbra zmcertmgr -fips deploycrt comm /tmp/tmp.usBeqCJNyV/0000_cert.pem /tmp/tmp.usBeqCJNyV/chain.pem
Unknown option: fips
```

I have already created a virtualmin server that I will use to migrate all the zimbra users to. But the cert expires tonight so I have to fix it or rush the migration.

Thanks!

**Comment by VojtechMyslivec** [comment]
If I understand it well, the `-fips` fix should be included in `8.8.15.p41`. So first make sure you have update the Zimbra to that patch. (And that parameter is used by `zmcertmgr` itself while calling OpenSSL, you don't have to edit the script and the `zmcertmgr` line).

However, I am not sure why it happened. No one other report that issue and I don't know why it stopped working

**Comment by dhcmega** [comment]
Hi @VojtechMyslivec thank you very much for the script and your time helping me.
I have already installed virtualmin in a new vps and moved all the accounts to it. SSL, dmarc, dkim, etc all working.

Since I started using Zimbra, it has been one problem after another. I even had to set the vps to auto-reboot every night to avoid sudden hangs.

Thanks to your script, I managed to extend Zimbra's usability by about two more years, but I've had enough of it.

Thanks again!
dhcmega
